# DSC 550 

### Week 1:  Exercise 1.2

Author: Kimberly Cable<br>
Date: 6-11-2022

In [12]:
#Load library
import pandas as pd
import numpy as np

**1. Load dataset**

In [13]:
try:
    video_games_df = pd.read_csv("Video_Games_Sales_as_at_22_Dec_2016.csv")
    
    video_games_df.info()
except Exception:
    print("There was an error reading the Video Games file")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16719 entries, 0 to 16718
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16717 non-null  object 
 1   Platform         16719 non-null  object 
 2   Year_of_Release  16450 non-null  float64
 3   Genre            16717 non-null  object 
 4   Publisher        16665 non-null  object 
 5   NA_Sales         16719 non-null  float64
 6   EU_Sales         16719 non-null  float64
 7   JP_Sales         16719 non-null  float64
 8   Other_Sales      16719 non-null  float64
 9   Global_Sales     16719 non-null  float64
 10  Critic_Score     8137 non-null   float64
 11  Critic_Count     8137 non-null   float64
 12  User_Score       10015 non-null  object 
 13  User_Count       7590 non-null   float64
 14  Developer        10096 non-null  object 
 15  Rating           9950 non-null   object 
dtypes: float64(9), object(7)
memory usage: 2.0+ MB


**2. Display the first 10 rows of the data.**

In [3]:
video_games_df.head(10)

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0  Nintendo     41.36     28.96      3.77         8.45         82.53   
1  Nintendo     29.08      3.58      6.81         0.77         40.24   
2  Nintendo     15.68     12.76      3.79         3.29         35.52   
3  Nintendo     15.61     10.93      3.28         2.95         32.77   
4  Nintendo     11.27      8.89     10.22         1.00         31.37   
5  Nintendo     23.20      2.26      4.22         0.58         30.26   
6  Nintendo     11.28      9.14      6.50         2.88         29.80   
7  Nintendo     13.96      9.18      2.93         2.84         28.92   
8  Nintendo     14.44      6.94      4.70         2.24         28.32   
9  Nintendo     26.93      0.63      0.28         0.47         28.31   

   Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0          76.0          51.0          8       322.0  Nintendo      E  
1           NaN           NaN        NaN         NaN       NaN    NaN  
2          82.0          73.0        8.3       709.0  Nintendo      E  
3          80.0          73.0          8       192.0  Nintendo      E  
4           NaN           NaN        NaN         NaN       NaN    NaN  
5           NaN           NaN        NaN         NaN       NaN    NaN  
6          89.0          65.0        8.5       431.0  Nintendo      E  
7          58.0          41.0        6.6       129.0  Nintendo      E  
8          87.0          80.0        8.4       594.0  Nintendo      E  
9           NaN           NaN        NaN         NaN       NaN    NaN

**3. Find the dimensions (number of rows and columns) in the data frame. What do these two numbers represent in the context of the data?**

In [4]:
video_games_df.shape

(16719, 16)

* There are 16,719 rows. Each row is information about a video game.
* There are 16 columns or features.  Features are data that describes each video game such as Name, Platform, Year it was Released, etc.

**4. Find the top five games by critic score.**

In [5]:
# Sort dataframe by Critic_Score in descending order to get the highest scores
top_five = video_games_df.sort_values(by=['Critic_Score'], ascending = False)
top_five.head(5)

Name Platform  Year_of_Release     Genre  \
227   Tony Hawk's Pro Skater 2       PS           2000.0    Sports   
57         Grand Theft Auto IV      PS3           2008.0    Action   
51         Grand Theft Auto IV     X360           2008.0    Action   
5350               SoulCalibur       DC           1999.0  Fighting   
165         Grand Theft Auto V     XOne           2014.0    Action   

                 Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
227             Activision      3.05      1.41      0.02         0.20   
57    Take-Two Interactive      4.76      3.69      0.44         1.61   
51    Take-Two Interactive      6.76      3.07      0.14         1.03   
5350    Namco Bandai Games      0.00      0.00      0.34         0.00   
165   Take-Two Interactive      2.81      2.19      0.00         0.47   

      Global_Sales  Critic_Score  Critic_Count User_Score  User_Count  \
227           4.68          98.0          19.0        7.7       299.0   
57           10.50          98.0          64.0        7.5      2833.0   
51           11.01          98.0          86.0        7.9      2951.0   
5350          0.34          98.0          24.0        8.8       200.0   
165           5.48          97.0          14.0        7.9       764.0   

                    Developer Rating  
227   Neversoft Entertainment      T  
57             Rockstar North      M  
51             Rockstar North      M  
5350                    Namco      T  
165            Rockstar North      M

**5. Find the number of video games in the data frame in each genre.**

In [6]:
# Get count of each Video game grouped by Genre
genre_count = video_games_df.groupby('Genre')['Name'].count()
genre_count

Genre
Action          3370
Adventure       1303
Fighting         849
Misc            1750
Platform         888
Puzzle           580
Racing          1249
Role-Playing    1500
Shooter         1323
Simulation       874
Sports          2348
Strategy         683
Name: Name, dtype: int64

**6. Find the first five games in the data frame on the SNES platform.**

In [7]:
snes_games = video_games_df[video_games_df['Platform'] == 'SNES']
snes_games.head()

Name Platform  Year_of_Release     Genre  \
18                      Super Mario World     SNES           1990.0  Platform   
56                  Super Mario All-Stars     SNES           1993.0  Platform   
71                    Donkey Kong Country     SNES           1994.0  Platform   
76                       Super Mario Kart     SNES           1992.0    Racing   
137  Street Fighter II: The World Warrior     SNES           1992.0  Fighting   

    Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
18   Nintendo     12.78      3.75      3.54         0.55         20.61   
56   Nintendo      5.99      2.15      2.12         0.29         10.55   
71   Nintendo      4.36      1.71      3.00         0.23          9.30   
76   Nintendo      3.54      1.24      3.81         0.18          8.76   
137    Capcom      2.47      0.83      2.87         0.12          6.30   

     Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
18            NaN           NaN        NaN         NaN       NaN    NaN  
56            NaN           NaN        NaN         NaN       NaN    NaN  
71            NaN           NaN        NaN         NaN       NaN    NaN  
76            NaN           NaN        NaN         NaN       NaN    NaN  
137           NaN           NaN        NaN         NaN       NaN    NaN

**7. Find the five publishers with the highest total global sales. Note: You will need to calculate the total global sales for each publisher to do this.**

In [8]:
# Group rows by Publisher and total Global_Sales per group
top_publisher_global_sales = video_games_df.groupby('Publisher', sort = False).agg({'Global_Sales': 'sum'})

# Sort Global Sales per Group in descending order and get top 5
top_publisher_global_sales.sort_values('Global_Sales', ascending = False).head(5)

Global_Sales
Publisher                                
Nintendo                          1788.81
Electronic Arts                   1116.96
Activision                         731.16
Sony Computer Entertainment        606.48
Ubisoft                            471.61

**8. Create a new column in the data frame that calculates the percentage of global sales from North America. Display the first five rows of the new data frame.**

In [9]:
def sales_percentage(regionSales, globalSales):
    '''
    Calculate percentage of regionalSales in globalSales
    
    args:
        regionSales(float)
        globalSales(float)
        
    return:
        percentage(float): percentage of regional sales in global sales
    '''
    try:
        return (regionSales / globalSales) * 100
    except ZeroDivisionError:
        return 0
    
# calcluate percentage of Global Sales from North America
video_games_df['NA_Global_Sales_Percent'] = sales_percentage(video_games_df['NA_Sales'], video_games_df['Global_Sales'])
video_games_df.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  \
0          51.0          8       322.0  Nintendo      E   
1           NaN        NaN         NaN       NaN    NaN   
2          73.0        8.3       709.0  Nintendo      E   
3          73.0          8       192.0  Nintendo      E   
4           NaN        NaN         NaN       NaN    NaN   

   NA_Global_Sales_Percent  
0                50.115110  
1                72.266402  
2                44.144144  
3                47.635032  
4                35.926044

**9. Find the number NaN entries (missing data values) in each column.**

In [10]:
video_games_df.isna().sum()

Name                          2
Platform                      0
Year_of_Release             269
Genre                         2
Publisher                    54
NA_Sales                      0
EU_Sales                      0
JP_Sales                      0
Other_Sales                   0
Global_Sales                  0
Critic_Score               8582
Critic_Count               8582
User_Score                 6704
User_Count                 9129
Developer                  6623
Rating                     6769
NA_Global_Sales_Percent       0
dtype: int64

**10. Try to calculate the median user score of all the video games. You will likely run into an error because some of the user score entries are a non-numerical string that cannot be converted to a float. Find and replace this string with NaN and then calculate the median. Then, replace all NaN entries in the user score column with the median value.**

In [11]:
# Initial attempt for Mean User Score
print(f"Mean User Score: {video_games_df['User_Score'].mean()}")

TypeError: can only concatenate str (not "int") to str

In [14]:
# View top 5 rows of User_Score
video_games_df['User_Score'].head()

0      8
1    NaN
2    8.3
3      8
4    NaN
Name: User_Score, dtype: object

In [15]:
# convert non-numeric values to NaN
video_games_df['User_Score'] = pd.to_numeric(video_games_df['User_Score'], errors = 'coerce')
video_games_df['User_Score'].head()

0    8.0
1    NaN
2    8.3
3    8.0
4    NaN
Name: User_Score, dtype: float64

In [16]:
# Convert NaN values to the median User_Score
video_games_df['User_Score'] = video_games_df['User_Score'].fillna(video_games_df['User_Score'].median())
video_games_df['User_Score'].head()

0    8.0
1    7.5
2    8.3
3    8.0
4    7.5
Name: User_Score, dtype: float64

In [18]:
# Calculate Mean User Score
print(f"Mean User Score: {video_games_df['User_Score'].mean()}")

Mean User Score: 7.329780489263756
